### PR TITLE
Property testing with quickcheck

### PR DESCRIPTION
**Pull request by snewt**
This PR represents an attempt to address issue #970. It also represents a portion of the meta issue for fuzzing #972.

The code base reflected here uses quickcheck to generate C headers that
include a variety of types including basic types, structs, unions,
function prototypes and function pointers. The headers generated by quickcheck
are passed to the `csmith-fuzzing/predicate.py` script. Examples of headers
generated by this iteration of the tooling can be viewed
[here](https://gist.github.com/snewt/03ce934f35c5b085807d2d5cf11d1d5c).

At the top of each header are two simple struct definitions,
`whitelistable` and `blacklistable`. Those types are present in the vector that
represents otherwise primitive types used to generate. They represent a naive
approach to exposing custom types without having to intuit generated type names like
`struct_21_8` though _any actual whitelisting logic isn't implemented
here_.

Test success is measured by the success of the
`csmith-fuzzing/predicate.py`
script. This means that for a test to pass the following must be true:
- bindgen doesn't panic
- the resulting bindings compile
- the resulting bindings layout tests pass

#### Usage
```bash
cd tests/property_test
cargo test
```

Some things I'm unsure of:
#### Where should this feature live?
At the moment it lives in `tests/property_test` but isn't run when
`cargo test` is invoked from bindgen's cargo manifest directory.

#### What's an acceptable ammount of time for these tests to take?
At this point, the source is genereated in ~1 second but the files are
large enough that it takes the `predicate.py` script ~30 seconds to run
through each one. In order for the tests to run in under a minute only 2 are
generated by quickcheck by default. This can be changed in the `test_bindgen`
function of the `tests/property_test/tests/fuzzed-c-headers.rs` file.

#### How do we expose the generated code for easy inspection?
For now the `run_predicate_script` function in the
`tests/property_test/tests/fuzzed-c-headers.rs` file contains a
commented block that will copy generated source in the `tests/property_test/tests`
directory. Should it be easier?

#### Special casing
There is some logic in the fuzzer that disallows 0 sized arrays because
tests will regulary fail due to issues documented in #684 and #1153. Should
this be special casing?

#### Does the fuzzer warrant its own crate?
After any iterations the reviewers are interested in required to make
this a functional testing tool, should/could the fuzzing library be made into
its own crate? I didn't move in that direction yet because having it all in one
place seemed like the best way to figure out what works an doesn't but I'm
interested in whether it might be useful as a standalone library.

#### What does it look like to expose more useful functionality?
I'm looking forward to feedback on how to make this a more useful tool
and one that provides the right configurability.

Thanks!

r? @fitzgen

### CODE REVIEW

**Comment by fitzgen** [comment]
@snewt thanks very much for this PR! And thanks for your patience -- I've been on vacation the last week, which is why you haven't heard back from me.

I'll try to take a look at the code and answer your open questions today, but it might get pushed back to tomorrow.

File: tests/property_test/tests/fuzzed-c-headers.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    //     .arg(&dir.path().to_str().unwrap())
+    //     .arg(&debug_output_path.to_str().unwrap())
+    //     .output()?;
+
+    Ok(Command::new(&predicate_script_path_string)
+        .arg(&header_path_string)
+        .output()?)
+
+    // omit close, from tempdir crate's docs:
+    // "Closing the directory is actually optional, as it would be done on drop."
```
Comment (fitzgen) [inline tests/property_test/tests/fuzzed-c-headers.rs]: Can remove this comment.

**Comment by snewt** [comment]
@fitzgen thanks for the awesome and thorough feedback, those are exactly sort of things I hoped / needed to hear. I'll keep you posted on progress or questions. Ha also I'm all about "nitpick-y" comments, I really appreciate getting familiar with and having the opportunity to live up to this project's code ideals!

**Comment by snewt** [comment]
__Address requested changes to quickchecking crate.__

- Remove `whitelistable` and `blacklistable` types.
- Rename test crate directory from `property_test` to `quickchecking`.
- Add new CI job that checks that this crate continues to build.
- Revise matching logic to be more idomatic.
- Phase out modular arithmetic in favor of `gen_range`.
- Incorporate `unreachable!` into match statements.
- Revise logic for accessing random element of vector, favor `choose` over `nth`.
- Proper punctuation and capitalization in comments. 
- Using actual structures rather than converting everything to strings in order to leverage type system.
- Add `#![deny(missing_docs)]` and filled in documentation required for the project to build again.
- Add special case logic so we don't generate structs with `long double` fields as it will cause tests to fail unitl issue #550 is resolved

Note on making sure we don't lose test cases we're interested in preserving: 
We're copying the directories `TempDir` makes so we get things like this:
```
├── bindgen_prop.1WYe3F5HZU1c
│   └── prop_test.h
├── bindgen_prop.H4SLI1JX0jd8
│   └── prop_test.h
```
I'm not sure that `TempDir` makes any claims about uniqueness, so collisions 
probably aren't impossible. I'm up for any suggestions on a more bulletproof 
solution.

_Tasks not addressed by this PR:_
* TODO: Add `cargo features` logic to allow generating problematic code.
* TODO: Make a [bin] target with CLI to manage test settings.
* TODO: Whitelisting and opaque types.
* TODO: Generate bitfields, C++, I-bogus-codegen cases.

Figured this would be a good point to update the PR but if any of the above TODO
items should be incorporated before moving forward I'm up for it!

Thanks for taking another look! 

r? @fitzgen

**Comment by snewt** [comment]
For reference, code that was generated after PR change requests were addressed is [here](https://gist.github.com/snewt/5a7517ea66778a4395ffbd6b23a036d1).

**Comment by fitzgen** [comment]
> I'm not sure that TempDir makes any claims about uniqueness, so collisions
> probably aren't impossible. I'm up for any suggestions on a more bulletproof
> solution.

This seems good enough to me.

**Comment by snewt** [comment]
@fitzgen ah awesome! yeah I can definitely file the follow up todos.

Thanks!!

**Pull request merged**